proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+- [@dotnet/macios](https://github.com/orgs/dotnet/teams/macios)
+- [@lambdageek](https://github.com/lambdageek)
+- [@SamMonoRT](https://github.com/SamMonoRT)
+- [@kotlarmilos](https://github.com/kotlarmilos)
+- [@JulieLeeMSFT](https://github.com/JulieLeeMSFT)
+- [@amanasifkhalid](https://github.com/amanasifkhalid)
+- [@davidortinau](https://github.com/davidortinau)
+- [@jkotas](https://github.com/jkotas)
+
+## Design
+
+### Runtime
+
+#### Calling Conventions
+
+Swift has two calling conventions at its core, the "Swift" calling convention, and the "SwiftAsync" calling convention. We will begin by focusing on the "Swift" calling convention, as it is the most common. The Swift calling convention has a few elements that we need to contend with. Primarily, it allows passing arguments in more registers. Additionally, it has two dedicated registers for specific components of the calling convention, the self register and the error register. The runtime support for this calling convention must support all of these features. The additional registers per argument is relatively easy to support, as each of our compilers must already have some support for this concept to run on Linux or Apple platforms today. We have a few options for the remaining two features we need to support.
+
+The "SwiftAsync" calling convention has an additional "async context" register and specific behaviors expected around how the Swift async state machine is called. Full investigation into the differences needed to implement it are still TBD.
+
+##### Self register
+
+We have two options for supporting the Self register in the calling convention:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+2. Specify that combining the `Swift` calling convention with the `MemberFunction` calling convention means that the first argument is the `self` argument.
+
+The first option seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables. The second option is a natural fit as the `MemberFunction` calling convention combined with the various C-based calling conventions specifies that there is a `this` argument as the first argument. Defining `Swift` + `MemberFunction` to imply/require the `self` argument is a perfect conceptual extension of the existing model.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use an attribute like `[SwiftError]` on a by-ref or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError<T>` on a by-ref or `out` parameter to indicate the error parameter.
+3. Use a special return type `SwiftReturn<T, TError>` to indicate the error parameter and combine it with the return value.
```

Comment:
Why do we need `TError` to be generic? I would expect that the low-level error to be a fixed pointer-sized type.

Reply:
I wasn't sure if the Swift error value could be passed by value and split across multiple registers (as can be done with the `self` register). If it can be passed by value, then we'd need this value to be generic to represent that. It looks like it can only be passed by pointer, so we can remove the generic argument.